docs/root/intro/arch_overview/data_sharing_between_filters.rst

```diff
@@ -0,0 +1,93 @@
+wants to convert it to some typed class object and act on it. Typed
+Metadata solves this problem by allowing Filters to register a one time
+conversion logic for a specific key. Incoming config metadata (via XDS) is
+converted to class objects at config load time. Filters can then obtain a
+typed variant of the metadata (usually protobuf) at runtime (per
+request/connection), thereby eliminating the need for filters to repeatedly
+convert from Protobuf::Struct to some internal object during
+request/connection processing.
+
+For example, if a filter that desires to have a convenience wrapper class
+over the opaque metadata with key "xxx.service.policy" in ClusterInfo, will
+register a ServicePolicyFactory that inherits from
+ClusterTypedMetadataFactory. The factory translates the protobuf.Struct
+into an instance of ServicePolicy class (inherited from
+FilterState::Object). When a Cluster is created, the associated
+ServicePolicy instance will be created and cached.
+
+Router:perFilterConfig
+----------------------
+
+In HTTP routes, per_filter_config is a special case of typed metadata where
+filters can  have route/virtualhost specific configuration in addition to a
+global filter config common to all routes. This configuration is converted
+and embedded into the route table. Its upto the filter implementation to
+treat the route-specific filter config as a replacement to global config or
+an enhancement. The fault filter uses this technique to provide per-route
+fault configuration.
+
+PerFilterConfig is a map<string, Struct>. The Connection manager iterates
+over this map and invokes the filter factory interface
+createRouteSpecificFilterConfig to parse/validate the struct value and
+convert it into a typed class object that’s stored with the route
+itself. Filters can then query the route-specific filter config during
+request processing.
+
+perFilterConfig shares same traits and design goals as the typed
+metadata. Its a specialized implementation focussing on per-route filter
+configuration (whether the filter implementation treats it as metadata or
+actual configuration is upto the filter author).
+
+
+Dynamic State
+=============
+
+Dynamic state is generated per network connection or per HTTP
+stream. Dynamic state can be mutable if desired by the filter generating
+the state.
+
+Envoy::Network::Connection and Envoy::Http::Filter provide a StreamInfo
+object that contains information about the current TCP connection and HTTP
+stream (i.e., HTTP request/response pair) respectively. StreamInfo contains
+a set of fixed attributes as part of the class definition (e.g., HTTP
+protocol, requested server name, etc.). In addition, it provides a facility
+to store typed/untyped objects (map<string,
+FilterState::Object>). FilterState objects can be either write-once, or
+write-many.
+
+*Note*: Existing dynamic metadata interface will be removed. Instead,
+filter authors can store the same protobuf structs under the read-write
+filterstate.
```

Comment:
Maybe just say it's deprecated, since we will need to keep it around for the standard release cycle.

Reply:
Matt wanted to remove it